File: fotoapparat/src/test/java/io/fotoapparat/result/PhotoResultTest.java
package io.fotoapparat.result;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.junit.MockitoJUnitRunner;

import java.io.File;

import io.fotoapparat.photo.BitmapPhoto;
import io.fotoapparat.photo.Photo;
import io.fotoapparat.result.transformer.BitmapPhotoTransformer;
import io.fotoapparat.result.transformer.SaveToFileTransformer;
import io.fotoapparat.test.ImmediateExecutor;

import static io.fotoapparat.test.TestUtils.immediateFuture;
import static junit.framework.Assert.assertNotNull;
import static junit.framework.Assert.assertSame;
import static org.mockito.ArgumentMatchers.isA;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.verify;

@RunWith(MockitoJUnitRunner.class)
public class PhotoResultTest {

    static final PendingResult<Photo> PENDING_RESULT = new PendingResult<>(
            immediateFuture(
                    Photo.empty()
            ),
            new ImmediateExecutor()
    );

    @Test
    public void toPendingResult() throws Exception {
        // Given
        PhotoResult photoResult = new PhotoResult(PENDING_RESULT);

        // When
        PendingResult<Photo> pendingResult = photoResult.toPendingResult();

        // Then
        assertSame(
                PENDING_RESULT,
                pendingResult
        );
    }

    @Test
    public void toBitmap() throws Exception {
        // Given
        PendingResult<Photo> pendingResult = spy(PENDING_RESULT);

        PhotoResult photoResult = new PhotoResult(pendingResult);

        // When
        PendingResult<BitmapPhoto> result = photoResult.toBitmap();

        // Then
        assertNotNull(result);

        verify(pendingResult).transform(
                isA(BitmapPhotoTransformer.class)
        );
    }

    @Test
    public void saveToFile() throws Exception {
        // Given
        PendingResult<Photo> pendingResult = spy(PENDING_RESULT);

        PhotoResult photoResult = new PhotoResult(pendingResult);

        // When
        PendingResult<?> result = photoResult.saveToFile(new File(""));

        // Then
        assertNotNull(result);

        verify(pendingResult).transform(
                isA(SaveToFileTransformer.class)
        );
    }
}

File: fotoapparat/src/test/java/io/fotoapparat/result/PhotoResultTest.kt
package io.fotoapparat.result

import io.fotoapparat.log.Logger
import io.fotoapparat.test.ImmediateExecutor
import io.fotoapparat.test.immediateFuture
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito.spy
import org.mockito.junit.MockitoJUnitRunner
import java.io.File
import kotlin.test.assertNotNull
import kotlin.test.assertSame

@RunWith(MockitoJUnitRunner::class)
class PhotoResultTest {

    @Mock
    lateinit var logger: Logger

    lateinit var pendingResult: PendingResult<Photo>

    @Before
    fun setUp() {
        pendingResult = PendingResult(
                immediateFuture(
                        Photo.empty()
                ),
                logger,
                ImmediateExecutor
        )
    }

    @Test
    fun `Convert to pending result`() {
        // Given
        val photoResult = PhotoResult(pendingResult)

        // When
        val result = photoResult.toPendingResult()

        // Then
        assertSame(
                expected = pendingResult,
                actual = result
        )
    }

    @Test
    fun `Convert to bitmap`() {
        // Given
        val pendingResult = spy(pendingResult)

        val photoResult = PhotoResult(pendingResult)

        // When
        val result = photoResult.toBitmap()

        // Then
        assertNotNull(result)
    }

    @Test
    fun `Save to file`() {
        // Given
        val pendingResult = spy(pendingResult)

        val photoResult = PhotoResult(pendingResult)

        // When
        val result = photoResult.saveToFile(File(""))

        // Then
        assertNotNull(result)
    }
}